### PR TITLE
fix: validate that nested and default tasks match `testNamePattern`

### LINKED ISSUE #539: When testNamePattern is set, tests nested in a describe block will still run even if the test name does not match
### Describe the bug

When I set an arbitrary `testNamePattern` such as `"this text will not match anything"`, then any test that is nested with describe will still run.

### Reproduction

- Go to `vitest/examples/basic`
- Change the `vite.config.ts` test configuration to contain `testNamePattern: "this should not match anything"`
- Run the tests using `test` or `test:run` scripts

Output looks like:

```
 RUN  /Users/asw/Development/vitest/examples/basic

 √ test/basic.test.ts (2)
 √ test/suite.test.ts (3)

Test Files  2 passed (2)
     Tests  3 passed | 2 skipped (5)
      Time  1.22s (in thread 5ms, 22813.95%)
```

Would expect 0 tests to pass and 5 to be skipped, but the tests in `test/suite.test.ts` are not skipped.

### System Info

```shell
System:
    OS: macOS 12.1
    CPU: (12) x64 Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
    Memory: 323.29 MB / 32.00 GB
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 14.18.1 - ~/.volta/tools/image/node/14.18.1/bin/node
    Yarn: 1.22.17 - ~/.volta/tools/image/yarn/1.22.17/bin/yarn
    npm: 6.14.15 - ~/.volta/tools/image/npm/6.14.15/bin/npm
  Browsers:
    Chrome: 97.0.4692.71
    Firefox: 96.0
    Safari: 15.2
  npmPackages:
    vite: latest => 2.7.12
    vitest: latest => 0.1.14
```

### Used Package Manager

yarn

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

### PR DESCRIPTION
**Pull request by DerYeger**
Closes #539

### CODE REVIEW

**Comment by DerYeger** [comment]
There's still some refactoring to do, as I still have some tasks incorrectly included.

File: packages/vitest/src/runtime/collect.ts [hunk 1]
```diff
@@ -97,11 +104,16 @@ function interpretTaskModes(suite: Suite, namePattern?: string | RegExp, onlyMod
...
+        interpretTaskModes(t, namePattern, onlyMode)
 
       // if all subtasks are skipped, marked as skip
       if (t.mode === 'run') {
-        if (t.tasks.every(i => i.mode !== 'run'))
+        if (t.tasks.every(i => i.mode !== 'run')) {
           t.mode = 'skip'
+          if (t.file)
+            t.file.mode = 'skip'
+        }
```
Comment (patak-dev) [inline packages/vitest/src/runtime/collect.ts:116]: If we have a nested suite where all subtasks are skipped, this check could mark a whole file as skip when there are tasks to run in other suites in the file. Maybe we could add an extra check out in line 74 to mark the file as skip, or refactor it out of the loop and add it at the end of interpretTaskMode

Comment (DerYeger) [reply packages/vitest/src/runtime/collect.ts:116]: Ah yes, totally! I'll take another look

Comment (DerYeger) [reply packages/vitest/src/runtime/collect.ts]: I pushed an update.

You were right, too many test files were skipped.
The workaround requires a check for (transitive) parent suites that are marked as `only`.

File: packages/vitest/src/runtime/collect.ts [hunk 2]
```diff
@@ -97,6 +105,8 @@ function interpretTaskModes(suite: Suite, namePattern?: string | RegExp, onlyMod
     else if (t.type === 'suite') {
       if (t.mode === 'skip')
         skipAllTasks(t)
+      else if (!interpreted)
+        interpretTaskModes(t, namePattern, onlyMode, isIncluded)
```
Comment (patak-dev) [inline packages/vitest/src/runtime/collect.ts:109]: Shouldn't this be `includeTasks`? Given that you want to always run `interpretTaskModes` looks better to do so here for both cases and remove the need for the `interpreted` boolean

File: packages/vitest/src/runtime/collect.ts [hunk 3]
```diff
@@ -75,19 +77,25 @@ export async function collectTests(paths: string[], config: ResolvedConfig) {
 /**
  * If any tasks been marked as `only`, mark all other tasks as `skip`.
  */
-function interpretTaskModes(suite: Suite, namePattern?: string | RegExp, onlyMode?: boolean) {
-  if (onlyMode === undefined)
+function interpretTaskModes(suite: Suite, namePattern?: string | RegExp, onlyMode?: boolean, isIncluded?: boolean) {
+  if (onlyMode === undefined) {
     onlyMode = someTasksAreOnly(suite)
+    isIncluded = isIncluded || suite.mode === 'only'
```
Comment (patak-dev) [inline packages/vitest/src/runtime/collect.ts:83]:
```suggestion
    isIncluded ||= suite.mode === 'only'
```

File: packages/vitest/src/runtime/collect.ts (hunk 2 above)
Comment (DerYeger) [reply packages/vitest/src/runtime/collect.ts]: I updated my changes

**Review by patak-dev — APPROVED** [review]

**Comment by patak-dev** [comment]
Thanks for the refactoring!

**Pull request merged**